### DISCUSSION TITLE
how to change defaultLocale dynamically without redirect to separate route

### DISCUSSION THREAD

**Discussion by satanworker** [vinissimus/next-translate#445] [Q&A]
Hey! First of all, thank you so much next-translate is wonderful! 
I just can't find in the examples how to change default locale without redirect? I have multiple websites and each of them can have only 1 locale but it can be changed, what is the best way to do so? Again thank you so much

**Comment by aralroca** [comment]
And wouldn't it be more useful for each language to have its own URL for better SEO?

Next-translate uses [Next.js i18n routing](https://nextjs.org/docs/advanced-features/i18n-routing) by default to manage translations. And you can only change the language based on a URL change.

Maybe you could do it if you don't use the webpack loader _(`loader:false` in the [configuration](https://github.com/vinissimus/next-translate/tree/1.0.1#3-configuration))_ and you add the [i18nProvider](https://github.com/vinissimus/next-translate/tree/1.0.1#i18nprovider) yourself and set the language as an internal state.

**Reply by jasonwilliams** [reply]
@aralroca that’s good to know. In our project, languages are set from an upstream API rather than URL (langs are persisted in a database for logged in users) which will be common for a lot of websites. So from what you said I figured I could just set `ctx.locale` in serverSideProps for this to work. (Or at least it did for me). I didnt think a solution was possible until I found this thread.

I wonder if this should be included in the front page docs for “how to change language”. Because the current solutions there didn’t really help for this scenario. I think it would help many users who use APIs for language settings.

**Reply by VinceBT-BG** [reply]
@jasonwilliams Can you give an example of what you did ? Did you find a better solution since ? Did yo have to do it for every page of the app ?
